ctru-rs/src/applets/swkbd.rs

```diff
@@ -311,6 +368,71 @@ impl SoftwareKeyboard {
         self.state.filter_flags = filters.bits();
     }
 
+    /// Configure a custom filtering function to validate the input.
+    ///
+    /// The callback function will return a callback result and the error message to display when the input is invalid.
+    ///
+    /// # Notes
+    ///
+    /// The filter callback will work only for the next time the keyboard is used. After using it once, it must be set again.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # fn main() {
+    /// #
+    /// use ctru::applets::swkbd::{SoftwareKeyboard, CallbackResult};
+    /// let mut keyboard = SoftwareKeyboard::default();
+    ///
+    /// keyboard.set_filter_callback(|text| {
+    ///     if text.contains("boo") {
+    ///         println!("Ah, you scared me!");
+    ///     }
+    ///
+    ///     (CallbackResult::Ok, None)
+    /// });
+    /// #
+    /// # }
+    pub fn set_filter_callback<F>(&mut self, callback: F)
+    where
+        F: FnOnce(&str) -> (CallbackResult, Option<CString>),
```

Comment:
Hmm, that's true. I'd like to keep the `CString` design as it more closely represents what's underneath (to avoid conversions on the heap), but the `&str -> String` design is obviously much cleaner. Tell me which one you prefer.

Reply:
Hmm, maybe for the sake of allocation we could do `&CStr -> Cow<CStr>`? Then if someone wants to do regular string operations they could use `CStr::to_string_lossy()` or other conversion functions, then convert back afterwards? This makes the user ensure their string is valid — if we did the conversion internally, we'd have to figure out what we want to do in the error case, and I think this way probably makes it more flexible for the user to handle errors how they want.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4c20c6b02b66c96b40d5806465ee5e2c

Reply:
the Cow<Cstr> approach is what i'd do too